### DISCUSSION TITLE
How to extract only a specific subset of features for real-time prediction?

### DISCUSSION THREAD

**Discussion by teobucci** [blue-yonder/tsfresh#1032] [Q&A] [answered]
Hi,

I'm in a task of timeseries classification, with the peculiarity that I need to implement the **prediction part in real-time** with the following logic: every 1 second I look at the past 2.5 seconds and classify them as 1 or 0. That gives me at most 1 second to perform every inference.

My time-series is 30 timesteps x 63 features.
When performing the training, I can use the default:
```python
X = extract_features(df.drop('target', axis=1), column_id='id', column_sort=None,
                     impute_function=impute,
                     n_jobs=1)
```
to extract a large number of features without any hurry, then I perform a feature selection with
```python
SelectFromModel(
        estimator=LogisticRegression(
            C=0.1,
            penalty='l1',
            solver='liblinear',
            max_iter=1000,
            random_state=42
        )
    )
```
and I'm able to get a very good performance with a `RandomForestClassifier` (0.93 AUC), however, if I try perform the same exact steps in the real-time scenario (extract all, feature select, random forest) it takes too much time due to the extraction part.

To me it looks pointless to extract all the features while I perfectly know which ones I need, they are also different features depending on which of the 63 time-series is being considered.

Is there a way to let the extractor know exactly which ones I need? Possibly being flexible without hardcoding everything.

**Comment by nils-braun** [comment, marked as answer]
Yes, this should indeed be possible! 
When extracting features, you can choose the extraction settings that define which feature calculators (using which settings) to use. You can choose one of the predefined settings - but you can definitely also create one particularly for your own needs. We have a convenience function https://tsfresh.readthedocs.io/en/latest/api/tsfresh.feature_extraction.html#tsfresh.feature_extraction.settings.from_columns for this. My suggestion would be:
1. extract all features like you described
2. do the selection
3. Compile the settings object using the from_columns method and store it (settinhs are pickleable)
4. For the real-time extraction, only set these settings. 
You might need to experiment a bit with the feature calculator. Some are particularly slow (the extremely slow are marked with high_comp_cost, if you do not want to extract them even in the beginning use the efficient parameters) so it might be worth having a bit less accuracy for your later task but faster execution time. 

You can read more on the things I described here: https://tsfresh.readthedocs.io/en/latest/text/feature_extraction_settings.html

**Reply by teobucci** [reply]
Thank you very much, Nils!
I had just stumbled upon that part of the doc late this morning and was experimenting with it. At first I didn't understand some things in the explanation. I will consider proposing a few changes to make it clearer as when I was researching I noticed a few Stack questions having trouble with it, if that might be appreciated!
Thanks again 😄
